ARCHITECTURE.md

```diff
@@ -0,0 +1,127 @@
+  generate `syntax_kinds` and `ast` modules, using `cargo gen-kinds` command.
+- `algo`: generic tree algorithms, including `walk` for O(1) stack
+  space tree traversal (this is cool) and `visit` for type-driven
+  visiting the nodes (this is double plus cool, if you understand how
+  `Visitor` works, you understand rust-analyzer).
+
+Test for ra_syntax are mostly data-driven: `tests/data/parser` contains a bunch of `.rs`
+(test vectors) and `.txt` files with corresponding syntax trees. During testing, we check
+`.rs` against `.txt`. If the `.txt` file is missing, it is created (this is how you update
+tests). Additionally, running `cargo gen-tests` will walk the grammar module and collect
+all `//test test_name` comments into files inside `tests/data` directory.
+
+See [#93](https://github.com/rust-analyzer/rust-analyzer/pull/93) for an example PR which
+fixes a bug in the grammar.
+
+
+### `crates/ra_editor`
+
+All IDE features which can be implemented if you only have access to a
+single file. `ra_editor` could be used to enhance editing of Rust code
+without the need to fiddle with build-systems, file
+synchronization and such.
+
+In a sense, `ra_editor` is just a bunch of pure functions which take a
+syntax tree as an input.
+
+The tests for `ra_editor` are `[cfg(test)] mod tests` unit-tests spread
+throughout its modules.
+
+### `crates/salsa`
+
+An implementation of red-green incremental compilation algorithm from
+rust compiler. It makes all rust-analyzer features on-demand. To be replaced
+with `salsa-rs/salsa` soon.
+
+
+### `crates/ra_analysis`
+
+A stateful library for analyzing many Rust files as they change.
+`AnalysisHost` is a mutable entity (clojure's atom) which holds
+current state, incorporates changes and handles out `Analysis` --- an
+immutable consistent snapshot of world state at a point in time, which
+actually powers analysis.
+
+
+### `crates/ra_lsp_server`
+
+An LSP implementation which uses `ra_analysis` for managing state and
+`ra_editor` for actually doing useful stuff.
+
+See [#79](https://github.com/rust-analyzer/rust-analyzer/pull/79/) as an
+example of PR which adds a new feature to `ra_editor` and exposes it
+to `ra_lsp_server`.
+
+
+### `crates/cli`
+
+A CLI interface to rust-analyzer.
+
+### `crate/tools`
+
+Code-gen tasks, used to develop rust-analyzer:
+
+- `cargo gen-kinds` -- generate `ast` and `syntax_kinds`
+- `cargo gen-tests` -- collect inline tests from grammar
+- `cargo install-code` -- build and install VS Code extension and server
+
+### `editors/code`
+
+VS Code plugin
+
+
+## Common workflows
+
+To try out VS Code extensions, run `cargo install-code`. To see logs from the language server,
+set `RUST_LOG=info` env variable. To see all communication between the server and the client, use
+`RUST_LOG=gen_lsp_server=debug` (will print quite a bit of stuff).
+
+To run tests, just `cargo test`.
+
+To work on VS Code extension, launch code inside `editors/code` and use `F5` to launch/debug.
```

Comment:
It may be worth mentioning that if you already have the extension installed and want to just hack on the language server can you do a `cargo install --path crates/ra_lsp_server --force`

Reply:
It might be worth adding a new cargo alias like `cargo update-server` for this.

Reply:
I would also suggest listing some useful `npm` commands you can use during development. I just created one to automatically apply all tslint and prettier suggestions, which can be called with `npm run fix` (see https://github.com/rust-analyzer/rust-analyzer/pull/115)